mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergPartitionTextParser.java

```diff
@@ -0,0 +1,111 @@
+ */
+class HiveIcebergPartitionTextParser {
+  private static final int PARTITION_TEXT_MAX_LENGTH = 1000;
+
+  private static final Pattern BUCKET_PATTERN = Pattern.compile("bucket\\((.*?),(.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern TRUNCATE_PATTERN = Pattern.compile(
+      "truncate\\((.*?),(.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern YEAR_PATTERN = Pattern.compile("year\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern MONTH_PATTERN = Pattern.compile("month\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern DAY_PATTERN = Pattern.compile("day\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern HOUR_PATTERN = Pattern.compile("hour\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern ALWAYS_NULL_PATTERN = Pattern.compile(
+      "alwaysNull\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final String INTEGER_PATTERN = "\\d+";
+  private static final String PIPE_DELIMITER = "\\|";
+
+  private HiveIcebergPartitionTextParser() {
+  }
+
+  private static boolean build2ArgTransform(Pattern transformPattern, BiConsumer<String, Integer> transformBuilder,
+                                            String trimmedPart, int intArgPos) {
+    Matcher matcher = transformPattern.matcher(trimmedPart);
+    if (matcher.find()) {
+      Preconditions.checkArgument(matcher.groupCount() == 2,
+          "Cannot parse 2-arg partition transform from text part: %s", trimmedPart);
+
+      String intArg = matcher.group(intArgPos).trim();
+      Preconditions.checkArgument(intArg.matches(INTEGER_PATTERN),
+          "Cannot find integer as argument %s in 2-arg partition transform: %s", intArgPos, trimmedPart);
+      int transformArg = Integer.parseInt(matcher.group(intArgPos).trim());
+
+      String columnName = matcher.group(intArgPos % 2 + 1).trim();
+      transformBuilder.accept(columnName, transformArg);
+      return true;
+    }
+
+    return false;
+  }
+
+  private static boolean build1ArgTransform(Pattern transformPattern, Consumer<String> transformBuilder,
+                                            String trimmedPart) {
+    Matcher matcher = transformPattern.matcher(trimmedPart);
+    if (matcher.find()) {
+      Preconditions.checkArgument(matcher.groupCount() == 1,
+          "Cannot parse 1-arg partition transform from text part: %s", trimmedPart);
+      String columnName = matcher.group(1).trim();
+      transformBuilder.accept(columnName);
+      return true;
+    }
+
+    return false;
+  }
+
+  public static PartitionSpec fromText(Schema schema, String text) {
+    Preconditions.checkArgument(text.length() < PARTITION_TEXT_MAX_LENGTH,
+        "Partition spec text too long: max allowed length %s, but got %s",
+        PARTITION_TEXT_MAX_LENGTH, text.length());
+    String[] parts = text.split(PIPE_DELIMITER);
```

Comment:
As @marton-bod pointed out in one of our discussions, there is a possibility to have columns with special characters in the column names, like `,|` etc. See: [HIVE-25222](https://issues.apache.org/jira/browse/HIVE-25222) - Fix reading Iceberg tables with a comma in column names.

Reply:
This could further complicate the parsing, and I am not sure it is worth it, but this is definitely something that the users should be aware.

Reply:
One other thing to consider is that users will be used to the `,` comma delimiter as the syntax from Spark SQL/Hive4/Impala, e.g. `partitioned by (bucket(id,16), category)`, so the pipe could feel unnatural.
I know this would complicate the parsing logic, but something to consider also.

Reply:
thanks for the suggestion, I have updated to allow an alternative delimiter, similar to the approach taken in HIVE-25222. 

Regarding using `,` instead of `|`, yes `|` is chosen mainly to avoid complexity and allow the use of `string.split` instead of complex parsing logic. Because this is not a long term solution, and other systems do not really have examples of parsing a delimited list of expressions, I think it is relatively reasonable to use `|`.  Please let me know if it is not enough.